skipchain/skipchain.go

```diff
@@ -37,44 +37,79 @@ const skipblocksID = "skipblocks"
 // Service handles adding new SkipBlocks
 type Service struct {
 	*onet.ServiceProcessor
-	// SkipBlocks points from SkipBlockID to SkipBlock but SkipBlockID is not a valid
-	// key-type for maps, so we need to cast it to string
-	*SkipBlockMap
-	Propagate messaging.PropagationFunc
-	gMutex    sync.Mutex
-	verifiers map[VerifierID]SkipBlockVerifier
-
-	// testVerify is set to true if a verification happened - only for testing
-	testVerify bool
+	// Sbm is the skipblock-map that holds all known skipblocks to this service.
+	Sbm           *SkipBlockMap
+	propagate     messaging.PropagationFunc
+	verifiers     map[VerifierID]SkipBlockVerifier
+	blockRequests map[string]chan *SkipBlock
+	lastSave      time.Time
 }
 
-// SkipBlockMap holds the map to the skipblocks so it can be marshaled.
-type SkipBlockMap struct {
-	SkipBlocks map[string]*SkipBlock
-}
-
-// ProposeSkipBlock takes a hash for the latest valid SkipBlock and a SkipBlock
+// StoreSkipBlock stores a new skipblock in the system. This can be either a
+// genesis-skipblock, that will create a new skipchain, or a new skipblock,
+// that will be added to an existing chain.
+//
+// The conode servicing the request needs to be part of the actual valid latest
+// skipblock, else it will fail.
+//
+// It takes a hash for the latest valid SkipBlock and a SkipBlock
 // that will be verified. If the verification returns true, the new SkipBlock
-// will be signed and added to the chain and returned.
-// If the the latest block given is nil it verify if we are actually creating
-// the first (genesis) block and creates it. If it is called with nil although
-// there already exist previous blocks, it will return an error.
-func (s *Service) ProposeSkipBlock(psbd *ProposeSkipBlock) (network.Message, onet.ClientError) {
-	prop := psbd.Proposed
+// will be stored, else it will be discarded and an error will be returned.
+//
+// If the latest block given is nil it will create a new skipchain and store
+// the given block as genesis-block.
+//
+// If the latest block is non-nil and exists, the skipblock will be added to the
+// skipchain after verification that it fits and no other block already has been
+// added.
+func (s *Service) StoreSkipBlock(psbd *StoreSkipBlock) (*StoreSkipBlockReply, onet.ClientError) {
+	prop := psbd.NewBlock
 	var prev *SkipBlock
+	var changed []*SkipBlock
 
-	if !psbd.LatestID.IsNull() {
+	if psbd.LatestID.IsNull() {
+		// A new chain is created, suppose all arguments in SkipBlock
+		// are correctly set up
+		prop.Index = 0
+		prop.Height = prop.MaximumHeight
+		prop.ForwardLink = make([]*BlockLink, 0)
+		// genesis block has a random back-link:
+		bl := random.Bytes(32, random.Stream)
```

Comment:
The random-backlink is used here to create a random hash. Without it two genesis-blocks with the same parameters on the same roster would have the same hash, and thus the same ID.
We can differentiate them by the `Index`, which is `0` for a genesis-block.